### PR TITLE
Add method Cell#cached?

### PR DESCRIPTION
**Pull request by PikachuEXE**
This is for logging cache hit rate for specific cells
Looking at caching server hit rate is useless and also inaccurate since rails 5.2's cache versioning

### CODE REVIEW

**Comment by PikachuEXE** [comment]
Spec fixed

**Comment by PikachuEXE** [comment]
Hello any more change required?

**Comment by PikachuEXE** [comment]
LOL almost forgot about this PR
Can this be merged?

**Comment by PikachuEXE** [comment]
Yo?

**Comment by PikachuEXE** [comment]
Hello?

**Comment by PikachuEXE** [comment]
@seuros :) ?
It's been sitting here *approved* for over 7 months
I think this boy deserves some attention 👋

**Comment by apotonick** [comment]
Do you have any example on where this method is used in your apps? :beers:

File: lib/cell/caching.rb [hunk 1]
```diff
@@ -59,6 +59,20 @@ def cache?(state, *args)
...
+    # It returns `false` without hitting cache store
+    #
+    # Same arguments as `#call`
+    def cached?(state=:show, *args)
+      return false unless cache?(state, *args)
+
+      key     = self.class.state_cache_key(state, self.class.version_procs[state].(self, *args))
+      options = self.class.cache_options[state].(self, *args)
+
+      cache_store.exist?(key, options)
```
[the remark this answers is not in the document]
Comment (PikachuEXE) [reply lib/cell/caching.rb]: It's from https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-exist-3F

File: test/cache_test.rb [hunk 2]
```diff
@@ -7,8 +7,17 @@ class CacheTest < Minitest::Spec
     def fetch(key, options, &block)
       self[key] || self[key] = yield
     end
+
+    def exist?(key, _options = nil)
+      !!self[key]
+    end
   end.new
 
+  def with_cache_cleanup
```
[the remark this answers is not in the document]
Comment (PikachuEXE) [reply test/cache_test.rb]: There is no `after(:each)` like rspec
So add this for cleanup after each test

File: lib/cell/caching.rb [hunk 3]
```diff
@@ -59,6 +59,20 @@ def cache?(state, *args)
...
 
+    # Method for getting whether the cell is cached
+    # If caching is disabled for the cell or the whole app
+    # It returns `false` without hitting cache store
+    #
+    # Same arguments as `#call`
+    def cached?(state=:show, *args)
+      return false unless cache?(state, *args)
+
+      key     = self.class.state_cache_key(state, self.class.version_procs[state].(self, *args))
```
[the remark this answers is not in the document]
Comment (PikachuEXE) [reply lib/cell/caching.rb]: Method extracted

**Comment by PikachuEXE** [comment]
I mainly use this method for 
- Tracking (whether transactions are having main page content cell cached)
- Page content "warming" (Keep important page content cells cached by running in bg workers)

**Comment by PikachuEXE** [comment]
Build failed in 2.2 due to
image

No test case is failing

**Comment by PikachuEXE** [comment]
Build for ruby 2.2 fixed
But I guess that change should be applied on `master` instead

**Comment by PikachuEXE** [comment]
Anything else I should respond to or update?

**Comment by PikachuEXE** [comment]
Hello again :3 ?

**Comment by apotonick** [comment]
Thanks @PikachuEXE for your contributions! :heart: We are still mulling over the decision whether or not this addition is really needed in the core code. It's a helpful method, but every method added is a method you need to maintain, test, and deprecate at some point. The more public API an object has, the worse it gets (see ActiveRecord for reference).

One idea I have is we could have a cells-cached gem and still merge your internal refactorings. With the download statistics we can then analyze how much this addition is being used - I am guessing, it won't be used a lot (#nojudging ;).

I apologize if that feels "unfair" but I'm planning a very slim Cells 5 and the more we add now, the more we have to deprecate, and in all fairness, there are more pressing issues than a caching tester, so please bear with me and think about my suggestions again.

**Comment by PikachuEXE** [comment]
I am totally fine with having an extension gem instead of merging it into core.
Tell me what I can/should do ;)

**Comment by PikachuEXE** [comment]
Also to let people know about your plan of releasing slim gem(s)
You might want to consider putting a `PULL_REQUEST_TEMPLATE` to let other guys know when they submit PR.

**Pull request closed without merging**